presto-main/src/main/java/com/facebook/presto/cost/CostCalculatorWithEstimatedExchanges.java

```diff
@@ -0,0 +1,194 @@
+                    .filter(node -> !node.isCoordinator())
+                    .count());
+        };
+    }
+
+    public CostCalculatorWithEstimatedExchanges(CostCalculator costCalculator, IntSupplier numberOfNodes)
+    {
+        this.costCalculator = requireNonNull(costCalculator, "costCalculator is null");
+        this.numberOfNodes = requireNonNull(numberOfNodes, "numberOfNodes is null");
+    }
+
+    @Override
+    public PlanNodeCostEstimate calculateCost(PlanNode node, StatsProvider stats, Lookup lookup, Session session, Map<Symbol, Type> types)
+    {
+        ExchangeCostEstimator exchangeCostEstimator = new ExchangeCostEstimator(numberOfNodes.getAsInt(), stats, lookup);
+        PlanNodeCostEstimate estimatedExchangeCost = node.accept(exchangeCostEstimator, null);
+        return costCalculator.calculateCost(node, stats, lookup, session, types).add(estimatedExchangeCost);
+    }
+
+    private static class ExchangeCostEstimator
+            extends PlanVisitor<PlanNodeCostEstimate, Void>
+    {
+        private final int numberOfNodes;
+        private final StatsProvider stats;
+        private final Lookup lookup;
+
+        ExchangeCostEstimator(int numberOfNodes, StatsProvider stats, Lookup lookup)
+        {
+            this.numberOfNodes = numberOfNodes;
+            this.stats = requireNonNull(stats, "stats is null");
+            this.lookup = requireNonNull(lookup, "lookup is null");
+        }
+
+        @Override
+        protected PlanNodeCostEstimate visitPlan(PlanNode node, Void context)
+        {
+            // TODO implement logic for other node types and return UNKNOWN_COST here (or throw)
+            return ZERO_COST;
+        }
+
+        @Override
+        public PlanNodeCostEstimate visitGroupReference(GroupReference node, Void context)
+        {
+            throw new UnsupportedOperationException();
+        }
+
+        @Override
+        public PlanNodeCostEstimate visitAggregation(AggregationNode node, Void context)
+        {
+            PlanNodeCostEstimate remoteRepartitionCost = CostCalculatorUsingExchanges.calculateExchangeCost(
+                    numberOfNodes,
+                    getStats(node.getSource()),
+                    REPARTITION,
+                    REMOTE);
+            PlanNodeCostEstimate localRepartitionCost = CostCalculatorUsingExchanges.calculateExchangeCost(
+                    numberOfNodes,
+                    getStats(node.getSource()),
+                    REPARTITION,
+                    LOCAL);
+            return remoteRepartitionCost.add(localRepartitionCost);
+        }
+
+        @Override
+        public PlanNodeCostEstimate visitJoin(JoinNode node, Void context)
+        {
+            return calculateJoinCost(
+                    node.getLeft(),
+                    node.getRight(),
+                    node.getDistributionType().orElse(JoinNode.DistributionType.PARTITIONED).equals(JoinNode.DistributionType.REPLICATED));
```

Comment:
This is much easier to read:
`node.getDistributionType().map(JoinNode.DistributionType.REPLICATED::equals).orElse(false)`
  
            
  Write

Reply:
Changed to
```
Objects.equals(node.getDistributionType(), Optional.of(JoinNode.DistributionType.REPLICATED))
```